MAINT: Start testing with Python 3.11.

felixxm opened the pull request:
We're having issues building `numpy` with Python 3.11 in Django tests, so I wanted to help make `numpy` compatible with Python 3.11.

felixxm committed:
MAINT: Start testing with Python 3.11.

mattip commented:
There are 3 warnings about `ob_shash`. One is from https://github.com/numpy/numpy/blob/28e624d1e78ee46d7040e6a099bda39d7fa60e3d/numpy/core/src/multiarray/scalarapi.c#L709

The others come from [cython](https://github.com/cython/cython/blob/c79ae55a7440459dae7667699284a72b09e89185/Cython/Utility/StringTools.c#L371). 
```
numpy/core/src/multiarray/scalarapi.c:709:13: warning: ‘ob_shash’ is deprecated [-Wdeprecated-declarations]
numpy/random/_mt19937.c:7144:13: warning: ‘ob_shash’ is deprecated [-Wdeprecated-declarations]
numpy/random/_mt19937.c:7145:13: warning: ‘ob_shash’ is deprecated [-Wdeprecated-declarations]
```

felixxm commented:
> There are 3 warnings about `ob_shash`. One is from
> 
> https://github.com/numpy/numpy/blob/28e624d1e78ee46d7040e6a099bda39d7fa60e3d/numpy/core/src/multiarray/scalarapi.c#L709

This looks unnecessary :thinking:, cached hash value for `obj` should not be computed yet.

mattip commented:
Thanks for getting this going. I think we should throw in a `if version_info.releaselevel == alpha` guard around the warnings check for now so we can produce a package that can be tested by other downstream consumers on the next cron build of NumPy. 

Then this will start failing on the beta releases if we have not cleared up the warnings. Does that make sense?

mattip commented:
... assuming tests pass after we can get a build

mattip commented:
Hmm. Now the [compile failed](https://github.com/numpy/numpy/runs/5865760667?check_suite_focus=true#step:4:1607) with another cython error, this one [has been fixed](https://github.com/cython/cython/pull/4667) and will be part of 0.29.29.
```
numpy/random/_mt19937.c:438:62: error: dereferencing pointer to incomplete type ‘PyFrameObject’ {aka ‘struct _frame’}
  438 |   #define __Pyx_PyFrame_SetLineNumber(frame, lineno)  (frame)->f_lineno = (lineno)
      |                                                              ^~
numpy/random/_mt19937.c:7855:5: note: in expansion of macro ‘__Pyx_PyFrame_SetLineNumber’
 7855 |     __Pyx_PyFrame_SetLineNumber(py_frame, py_line);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I wonder if there is an easy way to get CI to use Cython HEAD from a git commit for the 3.11 run?

felixxm commented:
Amazing, only one failure on Python 3.11 :+1:

felixxm committed:
BUG: Fix TestConversion.test_to_int_scalar() on Python.

The delegation of int() to __trunc__ was deprecated in Python 3.11, see
https://github.com/python/cpython/commit/b4bd1e1422997de61faf506b4916e83013bc7d21

mattip commented:
I could track the `ob_shash = 1` code back to [16 years ago](https://github.com/numpy/numpy/blame/cffb9714779adc0156c36bb55d0a1e8a29589408/numpy/core/src/arrayobject.c#L1347). Do we need to set the hash to -1 when we are sharing the data between the python object and the scalar?

mattip commented:
Looks like it is passing. Could you open an issue for updating to Cython 0.29.29, working around the `ob_shash` assignment, and reverting the `alpha` warning removal so we don't forget to update/fix/revert them?

charris commented:
I'd prefer to wait for a Cython release rather than complicate things. There are also Cython versions specified in `pyproject.toml` and `setup.py`. Looks like `setup.py` needs an update in any case. We normally start testing against Python with the beta releases.

h-vetinari commented:
> NumPy 1.23.x will be the first release to support 3.11. Note that we will be branching 1.22, in the middle of May, so the start of 1.23 development is about five weeks out.

Did you mean 1.24.x (because 1.22 has been released so the branching in May is going to be for 1.23...)?

If so, I'd like to ask that if the backports aren't too onerous, it would be nice to start supporting python 3.11 already with a 1.23 point-release (similar to how [1.19.3](https://github.com/numpy/numpy/releases/tag/v1.19.3) added support for python 3.9 and [1.21.3](https://github.com/numpy/numpy/releases/tag/v1.21.3) added support for 3.10)

charris commented:
> it would be nice to start supporting python 3.11 already with a 1.23 point-release

That is how we will do it once the first release candidate is out. We will start testing with the first beta, due 2022-05-06. And yes, it is 1.23 that will be branched next month.

neutrinoceros commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -8427,11 +8427,14 @@ def test_to_int_scalar(self):
+                    def __trunc__(self):
+                        return 3
+                assert_equal(3, int_func(np.array(HasTrunc())))
+                assert_equal(3, int_func(np.array([HasTrunc()])))
```
I would recommend adding an `else: pass` here, because that would enable the block to be automatically discovered and cleaned up with `pyupgrade` in the future (see https://github.com/asottile/pyupgrade#python2-and-old-python3x-blocks)

EwoutH replied:
@felixxm Could you implement this?

neutrinoceros replied:
your call. Pyupgrade doesn't need to be used continuously to be valuable, but I guess leaving these versioned branches around even after they become obsolete doesn't hurt so there's no strong incentive here.

felixxm replied:
Added

EwoutH commented on .github/workflows/build_test.yml:
```diff
@@ -60,7 +60,7 @@ jobs:
     strategy:
       matrix:
-        python-version: ["3.9", "3.10"]
+        python-version: ["3.9", "3.10", "3.11.0-alpha - 3.11.0"]
```
Maybe use `"3.11-dev"` instead of `"3.11.0-alpha - 3.11.0"` (see [docs](https://github.com/actions/setup-python#available-versions-of-python)).

charris replied:
The first beta is out. We should probably start providing nightly wheels once Cython 0.29.29 is out. I hope it is soon.

felixxm replied:
> The first beta is out. We should probably start providing nightly wheels once Cython 0.29.29 is out. I hope it is soon.

Yes :crossed_fingers:

felixxm replied:
Updated.

mattip commented on test_requirements.txt:
```diff
@@ -1,4 +1,4 @@
-cython>=0.29.28,<3.0
+-r cython_test_requirements.txt
```
Could you revert this and rebase this on top of main to get cython==0.29.29?

mattip commented on tools/travis-before-install.sh:
```diff
@@ -45,10 +45,10 @@ pip install --upgrade pip 'setuptools<49.2.0' wheel
+# A specific version of cython is required, so we use
+# cython_test_requirements.txt instead of simply writing
+# 'pip install setuptools wheel cython'.
+pip install -r cython_test_requirements.txt
```
Could you revert this change?

felixxm commented:
I also dropped ignoring deprecation warnings on Python's alphas.

mattip commented:
> It allows installing cython without grepping

Makes sense, even though it adds yet another layer of indirection to our requirements. I "resolved" my requests for changes. 

@rgommers should we have a pin on the cython version in the conda-facing `environment.yml`?

rgommers commented:
> > It allows installing cython without grepping
> 
> Makes sense, even though it adds yet another layer of indirection to our requirements. I "resolved" my requests for changes.

I don't quite understand this. Why can't we just the required versions in `pyproject.toml` and CI files directly? 

> @rgommers should we have a pin on the cython version in the conda-facing `environment.yml`?

Yes I think so, they should match the requirements for development with other types of environments like virtualenvs. If the minimum we need is `0.29.29`, it should say that I'd say. Upper bound of `<3.0` can be ignore though, because conda-forge won't package alpha/beta/rc versions of Cython.

h-vetinari commented:
> Upper bound of `<3.0` can be ignore though, because conda-forge won't package alpha/beta/rc versions of Cython.

Actually we [do](https://github.com/conda-forge/cython-feedstock/blob/dev/recipe/meta.yaml), but you won't ever get them by accident, only by adding a special channel (`-c conda-forge/label/cython_dev`).

EwoutH commented:
Now that #21543 is merged, could you rebase this PR? The Python 3.11 CI now might pass without any errors!

felixxm commented:
> Now that #21543 is merged, could you rebase this PR? The Python 3.11 CI now might pass without any errors!

Thanks :+1: , rebased.

mattip commented:
#21605 is merged but rerunning the failed CI job did not clear the error. Maybe a rebase on `main` is needed?

EwoutH commented:
@felixxm Could you rebase this PR?

felixxm commented:
Rebased

felixxm commented:
It works :rocket: :tada:

charris commented:
Could you get rid of `cython_test_requirements`. It isn't needed, the requirements are handled in the usual files.

felixxm commented:
> Could you get rid of `cython_test_requirements`. It isn't needed, the requirements are handled in the usual files.

Sure if you think it's not worth changing. Removed.

charris commented:
Thanks @felixxm .

EwoutH commented:
Could this PR also be backported to the 1.23 maintenance branch?

charris commented:
@EwoutH It is already marked for backport.
